--- old\regcomp.c	Fri Feb  2 01:17:18 2018
+++ new\regcomp.c	Fri Feb  2 01:17:18 2018
@@ -305,7 +305,11 @@
 {
   re_dfa_t *dfa = (re_dfa_t *) bufp->buffer;
   int node_cnt;
+#ifdef __GNUC__
   int icase = (dfa->mb_cur_max == 1 && (bufp->syntax & RE_ICASE));
+#else
+  bool icase = (dfa->mb_cur_max == 1 && (bufp->syntax & RE_ICASE));
+#endif
   for (node_cnt = 0; node_cnt < init_state->nodes.nelem; ++node_cnt)
     {
       int node = init_state->nodes.elems[node_cnt];
@@ -574,7 +578,11 @@
 static const bitset_t utf8_sb_map =
 {
   /* Set the first 128 bits.  */
+#ifdef __GNUC__
   [0 ... 0x80 / BITSET_WORD_BITS - 1] = BITSET_WORD_MAX
+#else
+  BITSET_WORD_MAX, BITSET_WORD_MAX, BITSET_WORD_MAX, BITSET_WORD_MAX
+#endif
 };
 #endif
 
@@ -920,9 +928,17 @@
 static void
 init_word_char (re_dfa_t *dfa)
 {
+#ifdef __GNUC__
   dfa->word_ops_used = 1;
   int i = 0;
   int ch = 0;
+#else
+  int i;
+  int ch;
+  dfa->word_ops_used = 1;
+  i = 0;
+  ch = 0;
+#endif
   if (BE (dfa->map_notascii == 0, 1))
     {
       /* Avoid uint32_t and uint64_t as some non-GCC platforms lack
@@ -959,9 +975,18 @@
 
  general_case:
   for (; i < BITSET_WORDS; ++i)
+#ifdef __GNUC__
     for (int j = 0; j < BITSET_WORD_BITS; ++j, ++ch)
       if (isalnum (ch) || ch == '_')
 	dfa->word_char[i] |= (bitset_word_t) 1 << j;
+#else
+  {
+    int j;
+    for (j = 0; j < BITSET_WORD_BITS; ++j, ++ch)
+      if (isalnum (ch) || ch == '_')
+        dfa->word_char[i] |= (bitset_word_t) 1 << j;
+  }
+#endif
 }
 
 /* Free the work area which are only used while compiling.  */
