--- old\regex.h	Sat Feb  1 20:52:50 2020
+++ new\regex.h	Sat Feb  1 20:52:50 2020
@@ -22,6 +22,11 @@
 
 #include <sys/types.h>
 
+#ifndef __GNUC__
+#define attribute_hidden
+#define libc_hidden_proto(name)
+#endif
+
 /* Allow the use in C++ code.  */
 #ifdef __cplusplus
 extern "C" {
@@ -614,7 +619,7 @@
    'configure' might #define 'restrict' to those words, so pick a
    different name.  */
 #ifndef _Restrict_
-# if defined __restrict || 2 < __GNUC__ + (95 <= __GNUC_MINOR__)
+# if defined __restrict || 2 < __GNUC__ + (95 <= __GNUC_MINOR__) || defined _MSC_VER
 #  define _Restrict_ __restrict
 # elif 199901L <= __STDC_VERSION__ || defined restrict
 #  define _Restrict_ restrict
