--- spal60sr\src\gui\FindNameUnit.original.cpp	Mon Jan 13 05:30:30 2003
+++ spal60sr\src\gui\FindNameUnit.cpp	Mon Jan 13 05:30:30 2003
@@ -7,6 +7,9 @@
 #include "MainUnit.h"
 #include "StringDivision.h"
 using namespace ssgCtrl;
+
+#define FIX_FIND_NAME 1
+
 //---------------------------------------------------------------------------
 #pragma package(smart_init)
 #pragma resource "*.dfm"
@@ -15,6 +18,13 @@
 __fastcall TFindNameForm::TFindNameForm(TComponent* Owner)
     : TForm(Owner)
 {
+#if FIX_GUI
+	::SetWindowLongA(Handle, GWL_STYLE, ::GetWindowLongA(Handle, GWL_STYLE) & ~WS_MINIMIZEBOX);
+#endif
+
+#if SNAP_WINDOW
+	AttachSnapWindow(Handle);
+#endif
 }
 //---------------------------------------------------------------------------
 //「初期化関数」
@@ -73,7 +83,12 @@
        	break;}
     case 1:{ //リスト検索
         TStrings *SList;
+#if !FIX_FIND_NAME
         switch(MainForm->GetSelectSubject()->GetType()){
+#else
+		TSSGSubject *selectSubject = MainForm->GetSelectSubject();
+		switch (selectSubject ? selectSubject->GetType() : stNONE) {
+#endif
         case stLIST: case stB_LIST: case stD_LIST:
             SList= MainForm->ListLBox->Items;
             break;
@@ -175,6 +190,7 @@
 //---------------------------------------------------------------------
 //大文字小文字半角全角カタカタひらがな非区別処理
 //---------------------------------------------------------------------
+#if !FIX_SORT
 string TFindNameForm::CnvString(string &Src)
 {
  char *tmpC = new char[Src.size()+1];
@@ -185,6 +201,26 @@
  delete[] tmpC;
  return tmpS;
 }
+#else
+EXTERN_C int __stdcall LCMapStringJISX0213(
+	IN           LCID   Locale,
+	IN           DWORD  dwMapFlags,
+	IN           LPCSTR lpSrcStr,
+	IN           int    cchSrc,
+	OUT OPTIONAL LPSTR  lpDestStr,
+	IN           int    cchDest);
+//---------------------------------------------------------------------
+string TFindNameForm::CnvString(string &Src)
+{
+	string Dest;
+	size_t cchSrc = Src.size();
+	size_t cchDest = cchSrc * 2;
+	Dest.resize(cchDest);
+	Dest.resize(LCMapStringJISX0213(findLCID, LCMAP_FULLWIDTH | LCMAP_KATAKANA | LCMAP_LOWERCASE,
+		Src.c_str(), cchSrc, Dest.begin(), cchDest));
+	return Dest;
+}
+#endif
 //---------------------------------------------------------------------------
 //「検索対象ラジオグループのクリックイベント」
 //---------------------------------------------------------------------------
