--- spal60sr\src\gui\ProcessAddUnit.original.dfm	Mon Nov 25 02:29:06 2002
+++ spal60sr\src\gui\ProcessAddUnit.dfm	Mon Nov 25 02:29:06 2002
@@ -75,7 +75,7 @@
       Top = 28
       Width = 129
       Height = 21
-      Caption = #31292#20685#12503#12525#12475#12473#12398#20877#25152#24471
+      Caption = #31292#20685#12503#12525#12475#12473#12398#20877#21462#24471
       TabOrder = 1
       OnClick = ReLoadBtnClick
     end
