--- spal60sr\src\base\ProcessAccessElement.original.cpp	Sat May 17 15:18:08 2003
+++ spal60sr\src\base\ProcessAccessElement.cpp	Sat May 17 15:18:08 2003
@@ -4,6 +4,14 @@
 #include <tlhelp32.h>
 #include "ProcessCtrl.h"
 #include "BitOperation.h"
+
+#define FIX_SPOILER_AL_61  1
+#define SHOW_ERROR_MESSAGE 1
+
+#if SHOW_ERROR_MESSAGE
+#include "ErrorMessage.h"
+#endif
+
 //---------------------------------------------------------------------
 //「引数の要素ベクタの処理サイズを求める関数」
 //---------------------------------------------------------------------
@@ -291,6 +299,9 @@
 	MaskedSrc	= TProcessAccessElementBase::Masking(data, mask);
 	tmpV		= TProcessAccessElementBase::Masking(tmpV, mask);
 
+#if FIX_SPOILER_AL_61
+#define data MaskedSrc
+#endif
 	//比較モードで振り分け(比較結果が真なら0、偽なら-1を返す。)
 	switch(compareType){
 	case prcsCtrl::ctEQUOL:		//完全一致
@@ -312,6 +323,9 @@
 		if(memcmp(data.begin(), tmpV.begin(), Size)<0) Ret = -1;
 		break;
 	}
+#if FIX_SPOILER_AL_61
+#undef data
+#endif
 	break;}
  default:
 	return 5;	//想定外のモード
