--- spal60sr\src\TSSGCtrl\SSBitList.original.h	Mon Mar  3 13:07:46 2003
+++ spal60sr\src\TSSGCtrl\SSBitList.h	Mon Mar  3 13:07:46 2003
@@ -10,11 +10,20 @@
 class TSSBitList : public TSSArgBoolVectorSubject
 {
 protected:
+#if !SUBJECT_STRING_TABLE
 	string lockStr;
+#else
+	CTableReferencedString lockStr;
+#endif
 
 	byte size;
+#if !SUBJECT_STRING_TABLE
 	string addressStr;	//アドレスを示す文字列
 	string indexFileName, realFileName;
+#else
+	CTableReferencedString addressStr;	// アドレスを示す文字列
+	CTableReferencedString indexFileName, realFileName;
+#endif
 	unsigned long maskByte;
 public:
 	TSSBitList(void) { type=ssgCtrl::stBITLIST; }
