--- spal60sr\src\TSSGCtrl\SSCalc.original.h	Sun Apr  6 21:46:30 2003
+++ spal60sr\src\TSSGCtrl\SSCalc.h	Sun Apr  6 21:46:30 2003
@@ -9,16 +9,27 @@
 using namespace std;
 //----------------------------------------------------------------------------
 
+#define FIX_DGRID 1
+
 class TSSCalc : public TSSGSubject
 {
 protected:
+#if !SUBJECT_STRING_TABLE
 	string lockStr;
+#else
+	CTableReferencedString lockStr;
+#endif
 
 	byte size;
 	long min, max;
 	bool isUnsigned, isBigEndian;
+#if !SUBJECT_STRING_TABLE
 	string addressStr;	//アドレスを示す文字列
 	string nowValHeadStr, nowValFootStr;
+#else
+	CTableReferencedString addressStr;	// アドレスを示す文字列
+	CTableReferencedString nowValHeadStr, nowValFootStr;
+#endif
 public:
 	TSSCalc(void) { type=ssgCtrl::stCALC; }
 	virtual ~TSSCalc(void) { }
