--- spal60sr\src\TSSGCtrl\SSCopy.original.h	Thu Jan 23 11:33:48 2003
+++ spal60sr\src\TSSGCtrl\SSCopy.h	Thu Jan 23 11:33:48 2003
@@ -11,11 +11,20 @@
 class TSSCopy : public TSSGSubject
 {
 protected:
+#if !SUBJECT_STRING_TABLE
 	string lockStr;
+#else
+	CTableReferencedString lockStr;
+#endif
 
 	unsigned long size;
+#if !SUBJECT_STRING_TABLE
 	string addressStr;	//アドレスを示す文字列
 	string srcAddressStr;
+#else
+	CTableReferencedString addressStr;	// アドレスを示す文字列
+	CTableReferencedString srcAddressStr;
+#endif
 public:
 	TSSCopy(void) { type=ssgCtrl::stCOPY; }
 	virtual ~TSSCopy(void) { }
