--- spal60sr\src\TSSGCtrl\SSCopy.original.h	Thu Jan 23 11:33:48 2003
+++ spal60sr\src\TSSGCtrl\SSCopy.h	Thu Jan 23 11:33:48 2003
@@ -11,23 +11,40 @@
 class TSSCopy : public TSSGSubject
 {
 protected:
+#if !LOCAL_MEMORY_SUPPORT
 	string lockStr;
+#else
+	CTableReferencedString lockStr;
+#endif
 
 	unsigned long size;
+#if !LOCAL_MEMORY_SUPPORT
 	string addressStr;	//アドレスを示す文字列
 	string srcAddressStr;
+#else
+	CTableReferencedString addressStr;	// アドレスを示す文字列
+	CTableReferencedString srcAddressStr;
+#endif
 public:
 	TSSCopy(void) { type=ssgCtrl::stCOPY; }
 	virtual ~TSSCopy(void) { }
 
 	void Setting(TSSGCtrl &SSGC);
     unsigned long Write(TSSGCtrl &SSGC, TSSArg *Arg);
+#if !SUBJECT_STRING_TABLE
 	string GetLockName(void){ return lockStr; };
+#else
+	string& GetLockName(void) const { return lockStr; }
+#endif
 	bool IsSameSubject(TSSGSubject *SSGS){ return (GetType()==SSGS->GetType()); }
 
 	//アクセサ
 	unsigned long GetSize(void){ return size; }
+#if !SUBJECT_STRING_TABLE
 	string GetAddressStr(void){ return addressStr; }
+#else
+	string& GetAddressStr(void) const { return addressStr; }
+#endif
 	string GetSrcAddressStr(void){ return srcAddressStr; }
 
 };
