--- spal60sr\src\TSSGCtrl\SSDir.original.cpp	Tue Mar 18 16:40:48 2003
+++ spal60sr\src\TSSGCtrl\SSDir.cpp	Tue Mar 18 16:40:48 2003
@@ -1,16 +1,33 @@
 #include "SSDir.h"
 
+#define ERRORSKIP_SUPPORT 1
+#define FIX_CLEARCHILD    1
+
+#if ERRORSKIP_SUPPORT
+#define AT_ERRORSKIP 0x2000
+#endif
+
 //---------------------------------------------------------------------
 //「下位項目の解放関数」
 //	・下位項目の解放を行います。
 //---------------------------------------------------------------------
 void TSSDir::ClearChild(void)
 {
+#if !FIX_CLEARCHILD
  vector<TSSGSubject *>::iterator 
 	it  = childVec.begin(),
 	End = childVec.end();
  for(; it!=End; it++) delete *it;
  childVec.clear();
+#else
+	for (vector<TSSGSubject *>::iterator it = childVec.begin(); it != childVec.end(); it++)
+	{
+		if ((*it)->GetType() == ssgCtrl::stDIR)
+			((TSSDir *)*it)->ClearChild();
+		delete *it;
+	}
+	vector<TSSGSubject *>().swap(childVec);
+#endif
 }
 //---------------------------------------------------------------------
 //「保持項目ツリーから、範囲指定付きで項目ベクタを引き抜く関数」
@@ -102,7 +119,12 @@
 	End = childVec.end();
  for(; VIt!=End; VIt++){
 	unsigned long Ret = (*VIt)->Write(SSGC, Arg);
+#if !ERRORSKIP_SUPPORT
 	if( Ret!=0 ) return Ret;
+#else
+	if (Ret != 0 && !SSGC.GetAttribute(this, AT_ERRORSKIP))
+		return Ret;
+#endif
  }
  return 0;
 }
