--- spal60sr\src\TSSGCtrl\SSDoubleToggle.original.h	Thu Jan 23 11:34:38 2003
+++ spal60sr\src\TSSGCtrl\SSDoubleToggle.h	Thu Jan 23 11:34:38 2003
@@ -12,12 +12,21 @@
 class TSSDoubleToggle : public TSSArgBoolVectorSubject
 {
 protected:
+#if !SUBJECT_STRING_TABLE
 	string lockStr;
+#else
+	CTableReferencedString lockStr;
+#endif
 	
 	bool isOffset;
 	vector<byte> statusVec;
+#if !SUBJECT_STRING_TABLE
 	string addressStr;	//アドレスを示す文字列
 	string indexFileName, dataFileName;
+#else
+	CTableReferencedString addressStr;	// アドレスを示す文字列
+	CTableReferencedString indexFileName, dataFileName;
+#endif
 public:
 	TSSDoubleToggle(void) { type=ssgCtrl::stD_TOGGLE; }
 	virtual ~TSSDoubleToggle(void) { }
