--- spal60sr\src\TSSGCtrl\SSDoubleList.original.h	Sun Mar  2 17:53:14 2003
+++ spal60sr\src\TSSGCtrl\SSDoubleList.h	Sun Mar  2 17:53:14 2003
@@ -11,13 +11,22 @@
 class TSSDoubleList : public TSSArgLongIndexSubject
 {
 protected:
+#if !SUBJECT_STRING_TABLE
 	string lockStr;
+#else
+	CTableReferencedString lockStr;
+#endif
 	
 	bool isOffset;
 	byte size;
 	vector<byte> statusVec;
+#if !SUBJECT_STRING_TABLE
 	string addressStr;	//アドレスを示す文字列
 	string indexFileName, dataFileName;
+#else
+	CTableReferencedString addressStr;	// アドレスを示す文字列
+	CTableReferencedString indexFileName, dataFileName;
+#endif
 public:
 	TSSDoubleList(void) { type=ssgCtrl::stD_LIST; }
 	virtual ~TSSDoubleList(void) { }
