--- spal60sr\src\TSSGCtrl\SSFloatCalc.original.h	Sun Apr  6 21:46:42 2003
+++ spal60sr\src\TSSGCtrl\SSFloatCalc.h	Sun Apr  6 21:46:42 2003
@@ -9,16 +9,27 @@
 using namespace std;
 //----------------------------------------------------------------------------
 
+#define FIX_DGRID 1
+
 class TSSFloatCalc : public TSSGSubject
 {
 protected:
+#if !SUBJECT_STRING_TABLE
 	string lockStr;
+#else
+	CTableReferencedString lockStr;
+#endif
 
 	byte size;
 	double min, max;
 	bool isBigEndian;
+#if !SUBJECT_STRING_TABLE
 	string addressStr;	//アドレスを示す文字列
 	string nowValHeadStr, nowValFootStr;
+#else
+	CTableReferencedString addressStr;	// アドレスを示す文字列
+	CTableReferencedString nowValHeadStr, nowValFootStr;
+#endif
 public:
 	TSSFloatCalc(void) { type=ssgCtrl::stFLOAT_CALC; }
 	virtual ~TSSFloatCalc(void) { }
