--- spal60sr\src\TSSGCtrl\SSFloatCalc.original.h	Sun Apr  6 21:46:42 2003
+++ spal60sr\src\TSSGCtrl\SSFloatCalc.h	Sun Apr  6 21:46:42 2003
@@ -9,6 +9,8 @@
 using namespace std;
 //----------------------------------------------------------------------------
 
+#define FIX_DGRID 1
+
 class TSSFloatCalc : public TSSGSubject
 {
 protected:
@@ -18,6 +20,9 @@
 	double min, max;
 	bool isBigEndian;
 	string addressStr;	//ƒAƒhƒŒƒX‚ðŽ¦‚·•¶Žš—ñ
+#if defined(FIX_DGRID) && FIX_DGRID
+public:
+#endif
 	string nowValHeadStr, nowValFootStr;
 public:
 	TSSFloatCalc(void) { type=ssgCtrl::stFLOAT_CALC; }
