--- spal60sr\src\TSSGCtrl\SSGAttributeElement.original.cpp	Mon May 26 13:24:04 2003
+++ spal60sr\src\TSSGCtrl\SSGAttributeElement.cpp	Mon May 26 13:24:04 2003
@@ -13,6 +13,10 @@
 #include "SSDoubleList.h"
 #include "SSList.h"
 #include "SSGAttributeElement.h"
+#if SCOPE_SUPPORT
+//---------------------------------------------------------------------
+int AttributeElementOrder = 0;
+#endif
 //---------------------------------------------------------------------
 //「TReplaceAttributeの初期化関数」
 //	・書式は、"オフセット値, SSCファイル名"。
