--- spal60sr\src\TSSGCtrl\SSList.original.h	Sun Mar  2 17:50:06 2003
+++ spal60sr\src\TSSGCtrl\SSList.h	Sun Mar  2 17:50:06 2003
@@ -11,12 +11,21 @@
 class TSSList : public TSSArgLongIndexSubject
 {
 protected:
+#if !SUBJECT_STRING_TABLE
 	string lockStr;
+#else
+	CTableReferencedString lockStr;
+#endif
 
 	byte size;
 	long firstVal, diffVal;
+#if !SUBJECT_STRING_TABLE
 	string addressStr;	//アドレスを示す文字列
 	string indexFileName;
+#else
+	CTableReferencedString addressStr;	// アドレスを示す文字列
+	CTableReferencedString indexFileName;
+#endif
 public:
 	TSSList(void) { type=ssgCtrl::stLIST; }
 	virtual ~TSSList(void) { }
