--- spal60sr\src\TSSGCtrl\SSSplit.original.cpp	Tue Mar 18 16:48:02 2003
+++ spal60sr\src\TSSGCtrl\SSSplit.cpp	Tue Mar 18 16:48:02 2003
@@ -11,7 +11,11 @@
  if(isSeted) return;
  
  vector<string> tmpV;
+#if !SUBJECT_STRING_TABLE
  SSGC.strD.List(code+SSGC.GetEndWith(this), ",", &tmpV, strD::etTRIM);
+#else
+ SSGC.strD.List((string&)code + SSGC.GetEndWith(this), ",", &tmpV, strD::etTRIM);
+#endif
  tmpV.resize(4);	//ãKíËêîÇ…
 
  if(tmpV[1]=="root"){
