--- spal60sr\src\TSSGCtrl\SSString.original.h	Sun Apr  6 21:46:52 2003
+++ spal60sr\src\TSSGCtrl\SSString.h	Sun Apr  6 21:46:52 2003
@@ -8,15 +8,32 @@
 using namespace std;
 //----------------------------------------------------------------------------
 
+#define UNICODE_SUPPORT 1
+
 class TSSString : public TSSGSubject
 {
 protected:
+#if !SUBJECT_STRING_TABLE
 	string lockStr;
+#else
+	CTableReferencedString lockStr;
+#endif
 
 	unsigned long size;
+#if UNICODE_SUPPORT
+public:
+	BOOLEAN isUnicode;
+protected:
+#endif
+#if !SUBJECT_STRING_TABLE
 	string addressStr;	//アドレスを示す文字列
 	string endWord;
 	string nowValHeadStr;
+#else
+	CTableReferencedString addressStr;	// アドレスを示す文字列
+	CTableReferencedString endWord;
+	CTableReferencedString nowValHeadStr;
+#endif
 public:
 	TSSString(void)  { type=ssgCtrl::stSTRING; }
 	virtual ~TSSString(void) {}
