--- spal60sr\src\TSSGCtrl\SSToggle.original.h	Sun Apr  6 21:46:56 2003
+++ spal60sr\src\TSSGCtrl\SSToggle.h	Sun Apr  6 21:46:56 2003
@@ -12,11 +12,19 @@
 class TSSToggle : public TSSGSubject
 {
 protected:
+#if !SUBJECT_STRING_TABLE
 	string lockStr;
-	
+
 	string addressStr;	//アドレスを示す文字列
 	string onCode, offCode;
 	string nowValHeadStr;
+#else
+	CTableReferencedString lockStr;
+
+	CTableReferencedString addressStr;	// アドレスを示す文字列
+	CTableReferencedString onCode, offCode;
+	CTableReferencedString nowValHeadStr;
+#endif
     unsigned long size;
 public:
 	TSSToggle(void) :size(0) { type=ssgCtrl::stTOGGLE; }
@@ -28,9 +36,9 @@
 	string GetLockName(void){ return lockStr; };
 	unsigned long GetArgType(void){ return ssgCtrl::atBOOL; }
 	bool IsSameSubject(TSSGSubject *SSGS){ return (GetType()==SSGS->GetType()); }
-	
+
 	string ToByteCode(TSSGCtrl &SSGC, TSSArg *Arg);
-	
+
 	//アクセサ
 	string GetAddressStr(void){ return addressStr; }
 	string GetOnCode(void){ return onCode; }
