--- spal60sr\src\TSSGCtrl\SSTrace.original.cpp	Sat Dec 14 19:12:00 2002
+++ spal60sr\src\TSSGCtrl\SSTrace.cpp	Sat Dec 14 19:12:00 2002
@@ -1,5 +1,7 @@
 #include "SSTrace.h"
 
+#define FIX_SSTRACE 1
+
 //---------------------------------------------------------------------
 //「初期設定関数」
 //
@@ -21,6 +23,9 @@
  lockStr		= tmpV[1];
 
  isSeted = true;
+#if FIX_SSTRACE
+ SSGC.SetLock(true, this, new TSSArgBool(true));
+#endif
 }
 //---------------------------------------------------------------------
 //「データ書き込み関数」
@@ -30,7 +35,9 @@
 //---------------------------------------------------------------------
 unsigned long TSSTrace::Write(TSSGCtrl &SSGC, TSSArg *Arg)
 {
+#if !FIX_SSTRACE
  Setting(SSGC);
+#endif
 
  TSmartHandle SHandle;
  if(( SHandle=SSGC.processCtrl.Open(PROCESS_VM_READ|PROCESS_VM_WRITE|PROCESS_VM_OPERATION) )==NULL)
